docs: add docs for enabling automated TLS and DNS for v3 on GCP (#3126)
jenkins-x/jx-docs: content/en/docs/v3/guides/infra/tls_dns.md

---
title: TLS and DNS
linktitle: TLS and DNS
description: Automated TLS and DNS
weight: 35
---

This section will describe how to enable automated TLS and DNS for your Jenkins X installation.

> To achive this we will use a couple of open source projects to help enable automated DNS for applications you 

Reviewer: ```suggestion
To achieve this we will use a couple of open source projects to help enable automated DNS for your applications.
```

For this guide we are going to assume you own a domain called `foo.io` which is managed by Google Cloud DNS, if it is not see [configure cloud dns to manage a domain](/docs/v3/guides/infra/google_cloud_dns).  

> Jenkins X services will have URLs like `https://hook-jx.bar.foo.io`.  The jx-requirement.yaml `namespaceSubDomain:` of `-jx` refers to the Kubernetes namespace the service is running in, this helps avoid clashes of the same application running in different namespaces in the same cluster.

Reviewer: I think the correct file name is jx-requirements.yml.

To start with we are focussed on GCP but will expand to other cloud providers.

# Google Cloud Platform

> ## Prerequisits

Reviewer: ```suggestion
## Prerequisites
```

- cluster created using Jenkins X [GCP Terraform getting started](/docs/v3/getting-started/gke/)
- own a domain, we will use [Google Domains](https://domains.google.com/registrar/) in this guide but any provider will work
- latest Jenkins X [upgrade](/docs/v3/guides/upgrade)

### Cloud Infrastructure
First we will configure the cloud infrastructure requirements:

- a GCP Service Account with the `dns.admin` role, see [here](https://cloud.google.com/iam/docs/understanding-roles#dns-roles) for more information
- a managed cloud dns zone, see [here](https://cloud.google.com/dns/docs/zones) for more information

To satisfy these requirements go to your infrastructure repository (contains Terraform main.tf) and add to your `values.auto.tfvars` the following:

```values.tf
parent_domain = "foo.io"
```

> Most people prefer to use a subdomain for a specific installation rather than purchasing one domain per cluster.  For example in a multi cluster setup you will probably want the same parent domain but each cluster using a differnt subdomain like development.foo.io, staging.foo.io and foo.io.

Reviewer: ```suggestion
Most people prefer to use a subdomain for a specific installation rather than purchasing one domain per cluster.  For example in a multi cluster setup you will probably want the same parent domain but each cluster using a different subdomain like development.foo.io, staging.foo.io and foo.io.
```

Reviewer: `foo.io` is a parent domain not a subdomain

To use a subdomain for this cluster add the following configuration:

```values.tf
subdomain     = "bar"
```

Now apply these changes:

> ```

Reviewer: ```suggestion
```bash
```

git add values.auto.tfvars
git commit -m 'feat: enable DNS cloud resources'
git push
```

> ```

Reviewer: ```suggestion
```bash
```

terraform plan
terraform apply
```

You can now see your managed zone in GCP [here](https://console.cloud.google.com/net-services/dns/zones)

### Cluster

Next we will configure the cluster requirements:

- Install [external-dns](https://github.com/kubernetes-sigs/external-dns#externaldns) - Kubernetes controller which watches for new Kubernetes Ingress resources and creates A records in Google Cloud DNS which will propogate globally across the internet

> - Install [cert-manager](https://cert-manager.io/docs/) - Kuberbetes controller which watches for requests to ask [Let's Encrypt](https://letsencrypt.org/) to issue a new wildcard TLS certificate for your domain and will manage this including renewals

Reviewer: ```suggestion
- Install [cert-manager](https://cert-manager.io/docs/) - Kubernetes controller which watches for requests to ask [Let's Encrypt](https://letsencrypt.org/) to issue a new wildcard TLS certificate for your domain and will manage this including renewals
```

To satisfy these requirements go to your cluster repository (contains helmfile.yaml)

Add external-dns to your clusters helmfile.yaml `releases` section:

> ```

Reviewer: ```suggestion
```yaml
```

- chart: bitnami/external-dns
```

Add cert-manager to your clusters helmfile.yaml `releases` section:

> ```

Reviewer: ```suggestion
```yaml
```

- chart: jetstack/cert-manager
```

Next we install
- a cluster wide [Issuer](https://cert-manager.io/docs/concepts/issuer/) which tells cert-manager how to validate you own your domain
- a namespaced [Certificate](https://cert-manager.io/docs/concepts/certificate/) to request a TLS certificate for services running in the `jx` namespace

> ```

Reviewer: ```suggestion
```yaml
```

- chart: jx3/acme
  name: tls-jx

>   values:

Reviewer 2: might want to add something like this to highlight the dependency on cert-manager so we avoid race conditions - chart: jetstack/cert-manager
  name: cert-manager
  namespace: jx
- chart: jx3/acme
  needs:
    - cert-manager
  name: tls-jx
  namespace: jx
  values:
    - issuer:
        enabled: true
        cluster: true

Author: thanks!  we might need to think about this one a little further because Jenkins X uses `helmfile template` with an output directory to write the generated kubernetes resources, it then does `kubectl apply` so I'm not totally sure if the 
```
needs:
  - cert-manager/cert-manager
```
will work as expected, there is a an option helmfile template flag
```
--output-dir-template value   go text template for generating the output directory. Default: {{ .OutputDir }}/{{ .State.BaseName }}-{{ .State.AbsPathSHA1 }}-{{ .Release.Name}}
```
which might be useful to see if we can split the CRDs into a different folder from the `ClusterIssuer` and then `kubectl` applies them in the right order?

i.e. 
```
helmfile template --output-dir=cluster --output-dir-template=[something]
```
generates folders:
```
/cluster
- /crds/foo.yaml
- /other cluster resources/bar.yaml
```
then 
```
kubectl apply -f cluster
```
first applies the CRDs and next the other cluster resources?

Reviewer 2: Think you're right! I think the two options are the one  you have suggested where we parse CRDs and modify their path so they are applied first. The only other one which feels like a worse  option is we support nested helmfiles and honour this structure in our `cluster` and `namespace` folders. I can't really think of much of a use case for  nested state  unless our chart list gets very large so for now probably your suggestion of separating and applying  CRDs before other resources makes the most sense. Sounds like an issue against `jx gitops` which I'll open later

  - issuer:
      enabled: true
      cluster: true
```

Cert-manager will use the cluster issuer to request a TLS certificate for each namespaces [Certificate](https://cert-manager.io/docs/concepts/certificate/) found.  The advantage here is that the same wildcard certificate is cached and reused for multiple namespaces reducing the risk of being [rate limited](https://letsencrypt.org/docs/rate-limits/) by Lets Encrypt.

Now install the `acme` chart into any namespace you require TLS, e.g.

> ```

Reviewer: ```suggestion
```yaml
```

- chart: jx3/acme
  name: tls-jx-staging
  namespace: jx-staging
- chart: jx3/acme
  name: tls-jx-production
  namespace: jx-production
```

> The domain from setting up your infrastructure in step one should appear in the `jx-requirements.yaml` of you cluster git repo.  Next configure your TLS options, update your `jx-requirements.yaml` with

Reviewer: ```suggestion
The domain from setting up your infrastructure in step one should appear in the `jx-requirements.yml` of you cluster git repo.  Next configure your TLS options, update your `jx-requirements.yml` with
```

> ```

Reviewer: ```suggestion
```yaml
```

ingress:
  domain: bar.foo.io
  externalDNS: false # this is unused and will be deprecated
  namespaceSubDomain: -jx.
  tls:
    email: "joe@gmail.com"
    enabled: true
    production: false
```

When first installing set `tls.production=false` so you use the Lets Encrtpt staging serivce which allows for more API calls before rate limniting requests.  They will issue a self-signed certificate so once happy everything is working change this to `tls.production=true`

Jenkins X uses a version stream to rollout tested versions of images, charts and default configuration.  The `jx-boot` job will apply these versions to your helmfile but you can also run the step yourself to see the defaults.

> ```

Reviewer: ```suggestion
```bash
```

jx gitops helmfile resolve
```

> ```

Reviewer: ```suggestion
```bash
```

git add values.auto.tfvars
git commit -m 'feat: enable DNS cloud resources'
git push
```

Now tail the admin logs and wait for the job to complete

> ```

Reviewer: ```suggestion
```bash
```
